platforms/core-runtime/service-registry-impl/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -472,13 +474,28 @@ public void stop() {
         public void add(SingletonService serviceProvider) {
             assertMutable();
             stoppable.add(serviceProvider);
+            checkDeclaredServices(serviceProvider.getDeclaredServiceTypes());
             collectProvidersForClassHierarchy(inspector, serviceProvider.getDeclaredServiceTypes(), serviceProvider);
             services.add(serviceProvider);
             for (AnnotatedServiceLifecycleHandler annotationHandler : lifecycleHandlers) {
                 notifyAnnotationHandler(annotationHandler, serviceProvider);
             }
         }
 
+        public void checkDeclaredServices(List<Class<?>> declaredServiceTypes) {
```

Comment:
Could this be expressed in an ArchUnit rule instead?

e.g., every method that's annotated with `@Provides` is not allowed to use these types?

Reply:
I haven't considered that option, as validation upon registration felt natural.

It could be expressed as an ArchUnit test, though I think it'd be less reliable. For instance, there is a case of registering services with procedural code (via `configure()` methods, calling `registration.add()`). Even if it was possible to cover some of those cases with ArchUnit tests, it wouldn't be comprehensive.

Granted, there is a much smaller chance of a developer writing something like `registration.add(Lazy.class, ServiceImpl.class)`, compared to using a generic as a return type of a `@Provides`-method. But I don't want to tie ArchUnit tests to all the ways our services framework allows to declare the services.

Why do you think ArchUnit is a better option?